components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt

```diff
@@ -24,27 +26,105 @@ class DefaultLoginValidationDelegate(
-                Result.CanBeCreated
+            val savedLoginByGuid = newLogin.guid?.let { storage.get(it) }
+            val savedLoginByElements = storage.getStoredLoginByImportantFields(newLogin)
+            // Check the merged login with the new password to display correct UI error states
+            val loginToCheck =
+                (savedLoginByGuid ?: savedLoginByElements)?.mergeWithLogin(newLogin)
+                    ?.copy(password = newLogin.password) ?: newLogin
+            val result = try {
+                storage.ensureValid(loginToCheck)
+                when {
+                    // GUIDs don't match, but everything else important matches so we can merge and update
+                    savedLoginByElements != null -> {
+                        Result.CanBeUpdated
+                    }
+                    newLogin.guid.isNullOrEmpty() || savedLoginByGuid == null -> {
+                        Result.CanBeCreated
+                    }
+                    newLogin.guid != savedLoginByGuid.guid -> {
+                        Result.CanBeCreated
+                    }
+                    // This means a password was saved for this site with a blank username. Update that record
+                    savedLoginByGuid.username.isEmpty() -> {
+                        Result.CanBeUpdated
+                    }
+                    // When a user name changes, we create a new login
+                    newLogin.username != savedLoginByGuid.username -> {
+                        Result.CanBeCreated
+                    }
+                    // Password has been updated
+                    else -> {
+                        Result.CanBeUpdated
+                    }
+                }
             } catch (e: InvalidRecordException) {
                 when (e.reason) {
-                    InvalidLoginReason.DUPLICATE_LOGIN -> Result.CanBeUpdated
+                    InvalidLoginReason.DUPLICATE_LOGIN -> Result.Error.Duplicate
                     InvalidLoginReason.EMPTY_PASSWORD -> Result.Error.EmptyPassword
                     InvalidLoginReason.EMPTY_ORIGIN -> Result.Error.GeckoError(e)
                     InvalidLoginReason.BOTH_TARGETS -> Result.Error.GeckoError(e)
                     InvalidLoginReason.NO_TARGET -> Result.Error.GeckoError(e)
-                    // TODO in what ways can the login fields be illegal? represent these in the UI
                     InvalidLoginReason.ILLEGAL_FIELD_VALUE -> Result.Error.GeckoError(e)
                 }
             }
+
+            result
+        }
+    }
+
+    companion object {
+        /**
+         * Searches storage to find a matching login to all A-S defined important elements
+         * (username, hostname, and formSubmitUrl or httpRealm)
+         * in case a user enters the same information as another login but doesn't match GUID
+         */
+        suspend fun LoginsStorage.getStoredLoginByImportantFields(newLogin: Login): Login? {
```

Comment:
Optional nit: I meant that I think this should probably be part of the `LoginsStorage` class itself, rather than an extension function.  I won't block on it though.